code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/GithubTokenHandler.java

```diff
@@ -19,100 +19,123 @@
+        List<GithubAccountInfo> orgAccountInfo = client.getOrgAccountInfo(token);
 
-        HttpResponse response = Request.Post(GITHUB_URL.get()).addHeader("Accept", "application/json").bodyForm(requestData).execute().returnResponse();
-        int statusCode = response.getStatusLine().getStatusCode();
-        if (statusCode != 200) {
-            throw new ClientVisibleException(statusCode);
+        if (StringUtils.isNotEmpty(userAccountInfo.getAccountId())) {
+            idList.add(userAccountInfo.getAccountId());
         }
-        jsonData = jsonMapper.readValue(response.getEntity().getContent());
 
-        if (jsonData.get("error") != null) {
-            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, (String) jsonData.get("error_description"));
+        for (GithubAccountInfo info : orgAccountInfo) {
+            if (StringUtils.isNotEmpty(info.getAccountId())) {
+                idList.add(info.getAccountId());
+                orgNames.add(info.getAccountName());
+            }
         }
 
-        AccountInfo accountInfo = getAccountInfo((String) jsonData.get("access_token"));
-        if (StringUtils.isNotEmpty(accountInfo.accountId)) {
-            jsonData.put("account_id", accountInfo.accountId);
+        if (SECURITY.get()) {
+            if (null == getWhitelistedUser(idList)) {
+                throw new ClientVisibleException(ResponseCodes.UNAUTHORIZED, "User is not whitelisted");
+            }
+            Account userAccount = authDao.getAccountByExternalId(userAccountInfo.getAccountId(), GITHUB_EXTERNAL_TYPE);
+            if (null == userAccount) {
+                authDao.createAccount(userAccountInfo.getAccountName(), GITHUB_USER_ACCOUNT_KIND, userAccountInfo.getAccountId(), GITHUB_EXTERNAL_TYPE);
+            }
+        } else {
+            authDao.updateAccountByKind(null, GITHUB_ADMIN_ACCOUNT_KIND, userAccountInfo.getAccountId(), GITHUB_EXTERNAL_TYPE);
         }
 
-        String accountName = accountInfo.accountName;
-        authDao.createAccount(accountName, GITHUB_ACCOUNT_KIND, accountInfo.accountId, GITHUB_EXTERNAL_TYPE);
-        return new Token(tokenService.generateEncryptedToken(jsonData, new Date(System.currentTimeMillis() + DAY_IN_MILLISECONDS)));
-    }
+        jsonData.put("account_id", userAccountInfo.getAccountId());
+        jsonData.put("whitelist", idList);
+        Date expiry = new Date(System.currentTimeMillis() + DAY_IN_MILLISECONDS);
 
-    private AccountInfo getAccountInfo(String token) throws IOException {
+        return new Token(tokenService.generateEncryptedToken(jsonData, expiry), userAccountInfo.getAccountName(), orgNames, null, null);
+    }
 
-        Map<String, Object> jsonData = null;
+    protected String getWhitelistedUser(List<String> idList) {
+        if(idList == null) {
+            return null;
+        }
+        List<String> whitelistedValues = fromCommaSeparatedString(WHITELISTED_ORGS.get());
```

Comment:
There a settings api for string list.  `io.cattle.platform.archaius.util.ArchaiusUtil.getList(String)`